Userland/Applications/FTPServer/main.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FTPServer.h"
+#include <AK/String.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/TCPServer.h>
+#include <LibCore/TCPSocket.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/Splitter.h>
+#include <LibGUI/TableView.h>
+#include <LibGUI/TextEditor.h>
+#include <LibGUI/Widget.h>
+#include <LibGUI/Window.h>
+#include <stdio.h>
+#include <unistd.h>
+
+int main(int argc, char** argv)
+{
+    int port = 21;
+    bool nogui = false;
+    bool allow_anonymous = true;
+    String root_directory = "/";
+
+    Core::ArgsParser args_parser;
+    args_parser.add_option(port, "Sets the port to use", "port", 'p', "port");
+    args_parser.add_option(nogui, "Runs in terminal-only mode", "nogui", 'n');
+    args_parser.add_option(allow_anonymous, "Allow anonymous logins", "anon", 'a');
+    args_parser.add_option(root_directory, "Set the root directory for anonymous logins", "root_directory", 'r', "root_directory");
+    args_parser.parse(argc, argv);
```

Comment:
I realise this is a draft, but so long as you're open to suggestions ...

It would be useful to have FTPServer load a config file. `/etc/FTPServer.conf` by default, but also allow a user-specified config file to be passed in as a command line argument so low privileged users can start their own FTP server if they want. This would allow specifying a bunch of reasonable default options. Alternatively, it could load the server configuration from a config file from the user's home directory automatically.

Also, it would be useful to be able to set the root directory for all users, not just the anonymous user. Limiting FTP to a default root directory, rather than the entire filesystem, is an extremely common use case, even in instances where operating system user accounts are used for authentication. Alternatively, a more complex solution would be allowing per-user root directories configured in a config file.

Reply:
> so low privileged users can start their own FTP server.

I cant listen/accept a socket without root, would `pledge` be a fix for that?

> It would be useful to have FTPServer load a config file. 

A settings file is on my todo list (hidden under `UI for settings`), so its something that I will get to eventually. That includes per-user root directories, and whatever else would be good as a setting

Reply:
> 
> 
> > so low privileged users can start their own FTP server.
> 
> I cant listen/accept a socket without root, would `pledge` be a fix for that?

That depends what error you're encountering.

Privileged ports <= 1024 will require `root` privileges. Low privileged users can bind to any other port (so long as it isn't in use).

If you're encountering `pledge` crashes, then yes you'll need to adjust your pledges. The crash should tell you which pledge was missing. Probably `unix accept inet`.

Reply:
Well, the ftp port is 21 so..

Reply:
> 
> 
> Well, the standard ftp port is 21, and thats a privileged port. but binding to it should be possible even for a non-root user imo shrug

There's a couple of things going on here.

In my opinion the best and easiest approach is to allow a non-root user to run the FTP server on a non-standard port. The user will be forced to specify a non-privileged port greater than 1024, such as `2121`.

You could try to set the server executable setuid root, bind to the socket, then drop privileges to the user's UID. But that approach is fraught with complexity and potential security issues.

Another alternative would be creating a dedicated `ftp` OS user for managing the service and have the executable run as this user.

Reply:
> I am not using pledges at the moment, so i was wondering if adding them would fix the issue.

Pledges can only ever further restrict functionality. see: [man 2 pledge](https://github.com/SerenityOS/serenity/blob/master/Base/usr/share/man/man2/pledge.md).

Eventually pledges should be added before this is merged. It is a fairly easy process. You can take a look at one of the other [Services](https://github.com/SerenityOS/serenity/tree/master/Userland/Services) for an example.

https://github.com/SerenityOS/serenity/blob/42d667645d750648949d18fe67fb09bf82d85a8c/Userland/Services/RequestServer/main.cpp#L18-L21

Reply:
Unprivileged port for non-root, pledges, and a settings file added 👍